elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore/pool.rb

```diff
@@ -0,0 +1,201 @@
+              logger.error "[#{e.class}] #{e.message} #{url}" if logger
+              raise ::Elasticsearch::Transport::Transport::HostUnreachableError.new(e, url), "Could not reach host #{e.class}: #{e.message}"
+            end
+
+            def update_urls(new_urls)
+              # Normalize URLs
+              new_urls = new_urls.map {|u| u.is_a?(URI) ? u : URI.parse(u) }
+
+              @state_mutex.synchronize do
+                # Add new connections
+                new_urls.each do |url|
+                  # URI objects don't have real hash equality! So, since this isn't perf sensitive we do a linear scan
+                  unless @url_info.keys.include?(url)
+                    logger.info("Elasticsearch pool adding node @ URL #{url}") if logger
+                    add_url(url)
+                  end
+                end
+
+                # Delete connections not in the new list
+                @url_info.each do |url,_|
+                  unless new_urls.include?(url)
+                    logger.info("Elasticsearch pool removing node @ URL #{url}") if logger
+                    remove_url(url)
+                  end
+                end
+              end
+            end
+
+            def add_url(url)
+              @url_info[url] ||= empty_url_meta
+            end
+
+            def remove_url(url)
+              @url_info.delete(url)
+            end
+
+            def empty_url_meta
+              {
+                :open => 0,
+                :dead => false
+              }
+            end
+
+            def with_connection
+              url, url_meta = get_connection
+
+              raise Error, "No Available connections!" unless url
+              yield url
+            rescue ::Elasticsearch::Transport::Transport::HostUnreachableError => e
+              mark_dead(url, e)
+              raise e
+            ensure
+              return_connection(url)
+            end
+
+            def mark_dead(url, error)
+              @state_mutex.synchronize do
+                url_meta = @url_info[url]
+                logger.warn("Marking url #{url} as dead! Last error: [#{error.class}] #{error.message}")
+                url_meta[:dead] = true
+                url_meta[:last_error] = error
+              end
+            end
+
+            def url_meta(url)
+              @state_mutex.synchronize do
+                @url_info[url]
+              end
+            end
+
+            def get_connection
+              @state_mutex.synchronize do
+                # The goal here is to pick a random connection from the least-in-use connections
+                # We want some randomness so that we don't hit the same node over and over, but
+                # we also want more 'fair' behavior in the event of high concurrency
+                raise "No available connections!" if @url_info.empty?
```

Comment:
There's quite a bit of an inconsistency in how we raise exceptions in this class. Sometimes we raise library specific errors (`ServerError`), sometimes we raise appropriate Ruby errors (`ArgumentError` in `normalize_host`), sometimes we raise generic `Error` and sometimes we "just raise" (`RuntimeError`), like in this example.

I don't think we need a finely-grained exception classes, but from the usability perspective, I'd like us to provide a nice way for people to `rescue` specific problems, and to see reasonable errors in their logs / NewRelic dashboards, and so on.